Circle restful (#2666)
AgoraIO/agora_doc_source: en-US/markdown/agora-chat/RESTful API Reference/Circle Management/manage-channels.md

# Channels

Circle is an online chat and community solution built on the Chat SDKs. It enables users to connect with communities of mutual interest and communicate with thousands of other users worldwide in real time.

Circle provides a three-layer structure, as follows:

- **Server**: Each server represents one of the communities within Circle. Users can search for and join servers by subject/interest. They can also create and manage their own servers.
- **Channel**: A server is organized into topic-based channels, where users can connect with other interested users to talk over specific topics. A default channel is automatically created during server creation.

> - **Thread**: Threads serve as temporary sub-channels inside an existing channel, to help better organize conversations in a busy channel.

Reviewer: ```suggestion
- **Thread**: Threads serve as sub-channels inside an existing channel, to help better organize conversations in a busy channel.
```

Reviewer: 跟产品经理沟通了一下，建议去掉 temporary。

| `description`     | String      |  The channel description. |
| `custom`       | String     |  The channel extension.   |
| `default_channel` | Number |  Whether the channel serves as the default one in the server:<li>`0`: No.</li><li>`1`: Yes.</li>  |
| `created`      | Number    |  The Unix timestamp (ms) when the channel was created.        |
| `server_id`     | String    |  The ID of the server to which the channel belongs.     |
| `channel_id`     | String    |  The channel ID.    |

## Authorization

Agora Chat RESTful APIs require Bearer HTTP authentication. Every time an HTTP request is sent, the following `Authorization` field must be filled in the request header:

```http
Authorization: Bearer ${YourAppToken}
```

In order to improve the security of the project, Agora uses a token (dynamic key) to authenticate users before they log in to the chat system. The Agora Chat RESTful API only supports authenticating users using app tokens. For details, see [Authentication using App Token](./generate_app_tokens?platform=RESTful).

## Create a channel

Creates a channel.

> Note that one server can have a maximum of 100 channels.

Reviewer: ```suggestion
Note that one server can have a maximum of 100 channels. To raise the limit, contact [support@agora.io](support@agora.io)。
```

| `invite_mode` | Number | Whether the channel invitation is automatically accepted:<li>`0`</li>: Yes.</li><li>`1`: No. After being invited, the invitee manually accepts or declines the invitation.</li> | No | 
| `description` | String | The channel description. The length cannot exceed 500 characters.  | No       | 
| `custom`  | String | The channel extension. The length cannot exceed 500 characters. | No       | 

### HTTP response
#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code. |
| `channel_id` | String | The channel ID. |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.
### Example
#### Request example

```shell
curl -X POST -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/channel'
-d '{
	"server_id" : "19VM9oPBasxxxxxx0tvWViEsdM",
	"name" : "chat",
	"type" : 0,

>   "rank"：0,

Reviewer: rank需缩进

}
```

## Modify a channel

Modifies the detailed information of the specified channel.

### HTTP request

```http
PUT https://{host}/{org_name}/{app_name}/circle/channel/{channel_id}?serverId={server_id}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Query parameter

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `server_id` | String | The ID of the server to which the channel belongs. | Yes  |

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Content-Type` | String | `application/json` | Yes  |
| `Accept` | String | `application/json` | Yes  |
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes |

#### Request body

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `name`        | String | The channel name. The length cannot exceed 50 characters.      | No   | 

> | `rank`        | Number | The maximum number of users in the channel:<li>(Default) `0`: 2,000; </li><li>`1`: 20,000</li><li>`2`: 100,000.    | No    | 

Reviewer: ```suggestion
| `rank`        | Number | The maximum number of users in the channel.<li>(Default) `0`: 2,000; </li><li>`1`: 20,000</li><li>`2`: 100,000.   <br/>To set this parameter to `1` or `2`, contact [support@agora.io](support@agora.io).   | No    | 
```

Reviewer: 若频道的最大成员设置为 20,000 或 100,000，需要联系 [support@agora.io](support@agora.io)。

Author: 既然要联系商务开通还把 1 和 2 写出来干嘛呢？

Reviewer: 告诉用户有这个选项，频道最大成员数可增加至 20,000或100,000。只不过得联系商务，需要单独收费的。

```http
GET https://{host}/{org_name}/{app_name}/circle/channel/public?serverId={server_id}&limit={limit}&cursor={cursor} 
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Query parameter

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `server_id` | String | The ID of the server to which the channel belongs. | Yes  |
| `limit` | Number    |   The number of channels to query per page. The value range is [1,20]. The default value is 20. This parameter is only required when retrieving by page.  |  No  |
| `cursor` | String  | The start position of the next query. This parameter is only required when retrieving by page.    |  No  |

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Accept`        | String | Yes | `application/json` |
| `Authorization` | String | Yes | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. |

### HTTP response

#### Response body

> If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:\

Reviewer: ```suggestion
If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:
```

| `server_id` | String | The ID of the server to which the channel belongs. | Yes  |

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Accept`        | String | `application/json`      | Yes       | 
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes       | 

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code. |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X DELETE -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/channel/XXX?serverId=XXX'
```

#### Response example

```json
{
    "code": 200
}
```

## Send a message

> Sending a message to a channel via RESTful API is basically the same as sending a message to a chat group. The only difference lies in the setting of `to`. For sending a channel message, you need to set the `to` parameter to the targeted channel ID.

Reviewer: ```suggestion
Sending a message to a channel via the RESTful API is basically the same as sending a message to a chat group. The only difference lies in the setting of `to`. For sending a channel message, you need to set the `to` parameter to the target channel ID.
```

Author: 我觉得这里用 target 是没问题的

For details, see [Send a group message](https://docs.agora.io/en/agora-chat/restful-api/message-management#send-a-group-message).

## Add a reaction

> Adds a reaction to the specified message.

Author: I agree. Wish we got a glossary. I add a link to https://docs.agora.io/en/agora-chat/client-api/reaction instead.

### HTTP request

```http
POST https://{host}/{org_name}/{app_name}/circle/reaction
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Content-Type` | String | `application/json` | Yes  |
| `Accept` | String | `application/json` | Yes  |
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes |

#### Request body

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `user_id`    | String | The user ID.    | Yes       |
| `message_id` | String | The ID of the message to which you want to add the reaction.   | Yes       |

> | `message` | String | The name of the reaction. The length cannot exceed 128 characters. | Yes       |

Reviewer: ```suggestion
| `message` | String | The ID of the emoji, same as that on the client. The maximum length is 128 bytes. | Yes       |
```

> | `message` | String | The name of the reaction. The length cannot exceed 128 characters. | Yes       |

Author: @haoxiuwen I have the same confusion. Could you please kindly pass this question?

> | `message` | String | The name of the reaction. The length cannot exceed 128 characters. | Yes       |

Reviewer: The ID of the emoji that is added as the reaction.
![image](https://user-images.githubusercontent.com/95393471/206646896-34d0af18-669d-4a91-8ccb-8dc3d6b11ebb.png)

| `code`  | Number    | The HTTP status code. |
| `count` | Number  | The number of reactions. |
| `msgId` | String  | The ID of the message to which the reactions belong.  |
| `reactionList` | List  | The list of the reactions and detailed information.  |
| `reactionList.reactionId` | String  | The reaction ID.  |
| `reactionList.message` | String  | The reaction name.  |
| `reactionList.state` | Bool  | Whether the current user sending this request has added a reaction to this message:<li>`true`: Yes.</li><li>`false`: No.</li> |
| `reactionList.count` | Number  | The number of users that have added this reaction to the message.  |
| `reactionList.userList` | Number  | The list of user IDs that have added this reaction.  |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X DELETE -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/reaction/user/XXX?msgIdList=XXX&channelId=XXX'
```

#### Response example

```json
{
    "code":200,
    "count" : 1,
    "reactions":[
        {

>             "msgId":"message id",

Author: Agree. @haoxiuwen Please evaluate if any change you should modify the source file.
![image](https://user-images.githubusercontent.com/101544891/206626927-b8ac77bd-30df-4c2e-9559-8ecd7ec173b7.png)

            "reactionList":[
                {

>                     "reactionId":"Reaction id",

Author: Agree. @haoxiuwen Please evaluate if any change you should modify the source file.
![image](https://user-images.githubusercontent.com/101544891/206626927-b8ac77bd-30df-4c2e-9559-8ecd7ec173b7.png)

>                     "reactionId":"Reaction id",

Reviewer: The value of the field ‘’reactionId‘’ can be changed to a specific figure, like 15890012560.

>                     "message":"Reaction name",

Author: Agree. @haoxiuwen Please evaluate if any change you should modify the source file.
![image](https://user-images.githubusercontent.com/101544891/206626927-b8ac77bd-30df-4c2e-9559-8ecd7ec173b7.png)

>                     "message":"Reaction name",

Reviewer: The value of the field `message` indicates the emoji ID.

>                     "state":"Whether the current user adds the reaction.",

Author: Agree. @haoxiuwen Please evaluate if any change you should modify the source file.
![image](https://user-images.githubusercontent.com/101544891/206626927-b8ac77bd-30df-4c2e-9559-8ecd7ec173b7.png)

                    "count":2,
                    "userList":[
                        "user1",
                        "user2"
                    ]
                }
            ]
        }
    ]
}
```

## Remove a reaction

Removes a reaction from the specified message.

### HTTP request

```http
DELETE https://{host}/{org_name}/{app_name}/circle/reaction/user/{user_id}?messageId={message_id}&message={message}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Query parameter

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `message_id` | String | The ID of the message to which the reactions belong. | Yes  |

> | `message` | String | The name of the reaction to remove. | Yes  |

Reviewer: ```suggestion
| `message` | String | The ID of the emoji that is added as the reaction. | Yes  |
```

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code. |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X POST -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/channel/XXX/user/remove?userId=XXX&serverId=XXX'
```

#### Response example

```json
{
    "code": 200
}
```

## Check whether a user exists in a channel

Checks whether a user exists in the specified channel.

### HTTP request

```http
GET https://{host}/{org_name}/{app_name}/circle/channel/{channel_id}/user/{user_id}?serverId={server_id}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

> 

Reviewer: 添加查询参数：
#### Query parameter
| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `server_id`        | String | The ID of the server to which the channel belongs.   | Yes       |

Author: 你在给出 source file 的时候有 review 过么？
![image](https://user-images.githubusercontent.com/101544891/206631244-745fedf7-8a42-4af1-9b0d-fa3b799e23f3.png)

Author: 还有这里
![image](https://user-images.githubusercontent.com/101544891/206633824-d3c24569-714b-49a0-863f-c1cb95b5a6b6.png)

| `limit` | Number    |   The number of members to query per page. The value range is [1,20]. The default value is 20. This parameter is only required when retrieving by page.  |  No  |
| `cursor` | String  | The start position of the next query. This parameter is only required when retrieving by page.  |  No  |

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Accept`        | String | `application/json`   | Yes       |
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes       |

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code.   |
| `count`   | Number     | The number of retrieved members.  |
| `users` | List | The list of users and detailed information. |
| `users.user_id` | String | The ID of the user. |

> | `users.role` | Number | The role of the user in the channel:<li>`0`</li>: Owner.<li>`1`</li>: Member. |

Author: @haoxiuwen Please evaluate if any change you should modify the source file. Maybe a value representing "admin/moderator" missing here?
![image](https://user-images.githubusercontent.com/101544891/206627399-59eb5e8f-935b-4f92-bee5-d7fa6ee62ce5.png)

> | `users.role` | Number | The role of the user in the channel:<li>`0`</li>: Owner.<li>`1`</li>: Member. |

Reviewer: `2` should be `1`. I will modify in the source file.

The server moderators are the moderators of all channels under the server.

> | `users.role` | Number | The role of the user in the channel:<li>`0`</li>: Owner.<li>`1`</li>: Member. |

Author: Thanks. I will add this line "The server moderators are the moderators of all channels under the server." for further explaination.

  "code" : 200
  "mute_users" : [
    {
      "expire" : 86400000,
      "user" : "u1"
    },
    {
      "expire" : 86400000,
      "user" : "u2"
    }
  ]
}
```

## Add a user to the channel mute list

Adds a user to the mute list of the specified channel.

### HTTP request

```http
POST https://{host}/{org_name}/{app_name}/circle/channel/{channel_id}/user/mute
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Content-Type` | String | `application/json` | Yes  |
| `Accept` | String | `application/json` | Yes  |
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes |

#### Request body

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `server_id`    | String | The ID of the server to which the channel belongs. | Yes       |
| `user_id`    | String | The ID of the user to be muted. | Yes       |

> | `duration`    | Number | The mute duration (ms). | Yes       |

Author: Good question. @haoxiuwen Any limitations for this param?
![image](https://user-images.githubusercontent.com/101544891/206627608-c81ffe3e-473e-43a8-99c3-5cfa16519753.png)

> | `duration`    | Number | The mute duration (ms). | Yes       |

Author: Thanks. Will add "If this parameter is not passed, the user is muted forever." to further explain how to use it.

  "name" :"thread-name"
}'
```

#### Response example

```json
{
  "code" : 200
}
```

### Retrieve the details of a thread

Retrieves the detailed information of the specified thread.

### HTTP request

```http
GET https://{host}/{org_name}/{app_name}/circle/thread/{thread_id}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Request header

| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `Accept`        | String | `application/json`  | Yes       |
| `Authorization` | String | The authentication token of the user or administrator, in the format of `Bearer ${YourAppToken}`, where `Bearer` is a fixed character, followed by an English space, and then the obtained token value. | Yes       |

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code.   |

> | `thread_id` | String | The thread ID.    |

Reviewer: ```suggestion
| `id` | String | The thread ID.    |
```

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code.   |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X DELETE -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/thread/XXX'
```

#### Response example

```json
{
    "code": 200
}
```

## Add a user to a thread

Adds users to the specified thread.

### HTTP request

```http
POST https://{host}/{org_name}/{app_name}/circle/thread/{thread_id}/user/join?userId={user_id}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

#### Request header

| Parameter | Type | Description | Required |

> | :----- | :----- | :------- | :-------- |

Reviewer: 补上 Content-Type：
| `Content-Type` | String | `application/json` | Yes |

Author: 你的 source file 也应该注意一下？
我这边理应只承担翻译的指责，或许你的源文件应该经过 peer review 确认无误了再给出呢？
![image](https://user-images.githubusercontent.com/101544891/206630140-4df492e6-bb51-4710-8477-361bc5723ed1.png)

### HTTP response

#### Response body

If the returned HTTP status code is `200`, the request succeeds, and the response body contains the following fields:

|  Parameter  |  Type  |  Description  |
| :-------- | :------- | :------- |
| `code`   | Number     | The HTTP status code.   |

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X POST -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/thread/XXX/user/join?userId=XXX'
```

#### Response example

```json
{
  "code" : 200
}
```

## Remove a user from a thread

Removes users from the specified thread.

### HTTP request

```http
POST https://{host}/{org_name}/{app_name}/circle/thread/{thread_id}/user/remove?userId={user_id}
```

#### Path parameter

For the parameters and detailed descriptions, see [Common parameters](#param).

> 

Reviewer: 添加查询参数

#### Query parameter
| Parameter | Type | Description | Required |
| :----- | :----- | :------- | :-------- |
| `user_id`  | String |  The user ID to remove from the thread.                        | Yes       |

Author: Ditto
![image](https://user-images.githubusercontent.com/101544891/206630705-eaeb56ea-143b-4a9a-b332-a9cfa96ad9df.png)

| `count`   | Number  | The number of retrieved threads.    |
| `threads` | List | The list of threads and their detailed information. |
| `threads.id` | String | The thread ID.    |
| `threads.msgId` | String | The ID of the message based on which the thread was created.   |
| `threads.channelId` | String | The ID of the channel to which the thread belongs. |
| `threads.owner`    | String | The ID of the user who creates the thread.     |
| `threads.created`    | Number | The Unix timestamp (md) when the thread was created.     |
| `cursor` | String  | The start position of the next query. This parameter is only required when retrieving by page.  |

For other parameters and detailed descriptions, see [Common parameters](#param).

If the returned HTTP status code is not `200`, the request fails. You can refer to [Status codes](#status-codes) for possible reasons.

### Example

#### Request example

```shell
curl -X GET -H 'Accept: application/json' -H 'Authorization: Bearer <YourAppToken>' 'http://XXX/XXX/XXX/circle/thread/joined?userId={user_id}&channelId=XXX&limit=1&cursor=XXX'
```

#### Response example

```json
{
  "code" : 200,

>   "threads" : [

Author: Nice spot. @haoxiuwen Take a look if any change you should modify the source file.
![image](https://user-images.githubusercontent.com/101544891/206628428-febb6c0f-564e-4ca7-af8e-331502c4cc76.png)

>   "threads" : [

Reviewer: The developer confirms that "count" is unnecessary.
